src/src/lib.rs

```diff
@@ -603,4 +595,88 @@ fn toRocReadError(err : std::io::Error) -> file_glue::ReadErr {
         // std::io::ErrorKind::InvalidFilename <- unstable language feature
         _ => file_glue::ReadErr::Unsupported,
     }
-}
\ No newline at end of file
+}
+
+#[no_mangle]
+pub extern "C" fn roc_fx_tcpConnect(host: &RocStr, port: u16) -> tcp_glue::ConnectResult {
+    match TcpStream::connect((host.as_str(), port)) {
+        Ok(stream) => {
+            let reader = BufReader::new(stream);
+            let ptr = Box::into_raw(Box::new(reader)) as u64;
+
+            tcp_glue::ConnectResult::Connected(ptr)
+        }
+        Err(err) => tcp_glue::ConnectResult::Error(to_tcp_connect_err(err)),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn roc_fx_tcpClose(stream_ptr: *mut BufReader<TcpStream>) {
+    unsafe {
+        drop(Box::from_raw(stream_ptr));
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn roc_fx_tcpRead(stream_ptr: *mut BufReader<TcpStream>) -> tcp_glue::ReadResult {
+    let reader = unsafe { &mut *stream_ptr };
+
+    match reader.fill_buf() {
+        Ok(received) => {
+            let received = received.to_vec();
+            reader.consume(received.len());
+
+            let rocList = RocList::from(&received[..]);
+            tcp_glue::ReadResult::Read(rocList)
+        }
+
+        Err(err) => tcp_glue::ReadResult::Error(to_tcp_stream_err(err)),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn roc_fx_tcpWrite(
+    msg: &RocList<u8>,
+    stream_ptr: *mut BufReader<TcpStream>,
+) -> tcp_glue::WriteResult {
+    let reader = unsafe { &mut *stream_ptr };
+    let mut stream = reader.get_ref();
+
+    match stream.write(msg.as_slice()) {
```

Comment:
I think you might want "write_all" here - otherwise, not all bytes may be written.

Reply:
Ah, right! Done